[MAINTENANCE] Refactor variable and rules configs in profiler into a single dict

cdkini opened the pull request:
Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- Combine the variables config and rules config into a profiler config dictionary.
- Refactored tests for Bobby & Alice to ensure continual coverage.


After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-


### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.


Thank you for submitting!

Commit:
refactor(profiler): Combine var and rules configs

The Profiler class originally took in two separate dictionaries:
  - variables config
  - rules config

To simplify the interface used by the end user, we've combined these values
into a single "profiler_config". Tests have been refactored to ensure
proper coverage after these changes.

cdkini commented on great_expectations/rule_based_profiler/profiler.py:
```diff
@@ -32,8 +32,7 @@ def __init__(
         *,
-        variables_configs: Optional[Dict[str, Dict]] = None,
-        rules_configs: Optional[Dict[str, Dict]] = None,
+        profiler_config: Dict[str, Dict[str, Dict]] = {},
```
I've gone ahead and made the default arg here an empty dictionary. My reasoning for this is that subsequent usages of `dict.get()` won't require unnecessary checks for `None` (since `NoneType` doesn't have a `get()` method). 

Might this have implications that I'm not considering?

talagluck replied:
Yes! I don't know exactly what happens, but you wind up with a Class attribute (?) that will change in unexpected ways if you re-instantiate the Profiler class. (I tried to do the exact same thing pretty recently, and this was explained to me by @anthonyburdi ).

cdkini replied:
Ah so keeping it as `None` is the recommended approach? I'll have to add checks in the method to ensure we're not using `get()` with a `NoneType` then.

@anthonyburdi would you mind telling me how that all works (or pointing me to a resource)? Would love to see what's going on behind the scenes :)

talagluck replied:
Yup! I think we'll need to change this to None.

This provides a bit more context, I think: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

anthonyburdi replied:
Great link Tal!
When using mutable types as defaults for constructors in python you end up with a single instance (sort of "class" variable) that can get modified by each instance (they all use a reference to the same dict) rather than what you want which is an instance variable tied to the specific instance. It leads to tricky bugs.
Also when changing back to `None` don't forget to mark the type as `Optional`

cdkini replied:
Really good to know, thank you both! Just when I thought I knew Python 🐍

cdkini commented on great_expectations/rule_based_profiler/profiler.py:
```diff
@@ -53,8 +52,10 @@ def __init__(
 
         self._rules = []
 
+        rules_configs: Dict[str, Dict] = profiler_config.get("rules", {})
```
Logic is the same as above for this and `variables_config`. Makes sense to return an empty dict if we don't have part of the config but I could see an argument made for the default `None`.

talagluck replied:
I think this is good for now (it will save us a bit of error handling so we don't accidentally try and loop over a None), though we may want to change this in the future if we decide to raise an error if the config is empty.

cdkini replied:
Do you think it makes sense to leave a TODO or is it okay as is?

talagluck replied:
OK as is - it doesn't feel like a must-do.

cdkini commented on tests/rule_based_profiler/test_profiler_user_workflows.py:
```diff
@@ -250,12 +250,10 @@ def test_alice_profiler_user_workflow_single_batch(
     data_context: DataContext = alice_columnar_table_single_batch_context
     # Load profiler configs & loop (run tests for each one)
-    profiler_config: str = alice_columnar_table_single_batch["profiler_config"]
+    yaml_config: str = alice_columnar_table_single_batch["profiler_config"]
```
This was originally named `profiler_config`, which conflicts with the new constructor arg for `Profiler`. Let me know if we can be more descriptive here or if this suffices.

talagluck replied:
Could you explain this a bit more? It looks like we're able to use the `profiler_config` name with no problem e.g. on L273 and L385. Why do we need to rename this?

cdkini replied:
Per the changes made in this PR, the new arg in the `Profiler` constructor is called `profiler_config` (which is of type `dict`). Prior to the changes made, `profiler_config` on L253 was a `str` that was ultimately used to generate the appropriate `dict` passed into the constructor.

I've renamed `full_profiler_config_dict` to just `profiler_config` on L256 so that the parameter and argument names match (i.e. `validator=validator`). Let me know if that clarifies things

talagluck replied:
Ah I see - makes sense!

cdkini commented:
Note that changes made here will cause merge conflicts with https://github.com/great-expectations/great_expectations/pull/2912.

talagluck reviewed (commented):
Looks good overall! Just a few comments/updates. Also, I don't think is specifically [MAINTENANCE] - it's not ideal, but [FEATURE] might be better.

cdkini commented:
@talagluck thanks for the review! I agree that MAINTENANCE is probably not idea; should I go ahead and make a new PR with an updated branch?

talagluck reviewed (commented):
Added some more comments! I don't think a new PR with a new branch is necessary, just changing the description for this one should be fine.

Commit:
refactor(profiler): Update constructor default val

Changed profiler args default val from an empty dict to None.

Per discussion with @anthonyburdi and @talagluck, mutable types as
defaults for constructors create unwanted class attrs.

Further reading: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

cdkini commented:
Made updates in response to last round of review notes; pending merge until other PR's in `rules_based_profiler` are complete.

cdkini commented:
Converting back to MAINTENANCE per discussion noted here: https://superconductive.atlassian.net/wiki/spaces/SUP/pages/86278279/Git+Repository+Best+Practices?focusedCommentId=186613761#comment-186613761

talagluck commented:
> Converting back to MAINTENANCE per discussion noted here: https://superconductive.atlassian.net/wiki/spaces/SUP/pages/86278279/Git+Repository+Best+Practices?focusedCommentId=186613761#comment-186613761

Interesting - my impression was that MAINTENANCE was specifically about meta stuff, like changes to our Azure pipeline configs.

talagluck commented:
Thanks for following up! I'll update my mental model

Commit:
Merge remote-tracking branch 'origin/develop' into MAINTENANCE/GE-160/GE-299/refactor_profiler_config_args

Commit:
Merge branch 'develop' into MAINTENANCE/GE-160/GE-299/refactor_profiler_config_args

Commit:
test(profiler): Remove validator arg from tests

cdkini commented:
b1b4b83 did not account for outdated use of `validator` in `Profiler` constructors in `test_profiler_user_workflows.py`. 0c5c8ef reflects these changes and passes test suite (at least on my machine).

alexsherstinsky reviewed (changes requested):
Team:

I would like to suggest an alternative approach:
1. We pull `variables` out of the `Rule` class -- because `variables` must be at the highest level, applicable to ALL rules.  Hence, `variables` should be be passed to the `Rule` constructor; they should be accessible to all `Rules` through the `profiler` reference, which would be passed as part of `runtime_environment` upon instantiation (using `instantiate_class_from_config()`).  This will entail small change in the rest of rule-based profiler component's implementation.
2. Big leap: the *entire* profiler is instantiated from within `DataContext` -- for example, the YML file, such as `tests/rule_based_profiler/bobby_user_workflow_verbose_profiler_config.yml`, would be processed all at once by reading it into a `CommentedMap` and using that to instantiate `Profiler` (again, using `instantiate_class_from_config()`).

Once this is working, it will position us for immediately offering a solution to the SaaS Pod, while in parallel moving on to add schema validation for our new profiler's configuration (using Marshmallow).  Thus, in many important ways, the rule-based profiler will behave analogously to `Checkpoints` and others -- in fact, we should also create a store for Profiler configuration YML files (just like we have a Checkpoint store for Checkpoint configurations).

This sounds like a lot, but it can be done in small and clean clean value-add increments quite straightforwardly.  Having said that, it might make sense to consider this work to be the focus of a separate epic, potentially the very next one.

Thanks.

alexsherstinsky reviewed (approved):
LGTM
